cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc

```diff
@@ -0,0 +1,833 @@
+
+. `MATCH (a)-[]->(b) RETURN a AS x, x.x + count(b.y)`
+* The sub-expression `x` is not a grouping key; it is the alias of a grouping key, which are not visible to <ProjectionItem>s within the same clause.
+
+. `MATCH (a)-[]->(b) RETURN a AS x, size(keys(x)) + count(b.y)`
+* The sub-expression `x` is not a grouping key; it is the alias of a grouping key, which are not visible to <ProjectionItem>s within the same clause.
+
+. `MATCH (a)-[c]->(b) WITH a, c, {dim: properties(b)} AS b RETURN a, b.dim.x, sum(c.p) + b.dim.x`
+* The sub-expression `b.dim.x` is not a recognized grouping key.
+
+=== Valid orderings
+
+The following clauses exhibit valid row orderings.
+For each example we list why it is valid.
+
+. `RETURN 1 + count(\*) AS x ORDER BY x`
+* The expression `x` is a recognized alias.
+
+. `RETURN 1, 1 + count(*) ORDER BY 2`
+* The expression `1` is a constant.
+
+. `RETURN $x + count($x) ORDER BY $x`
+* The expression `$x` is a parameter.
+
+. `MATCH (a) RETURN a.x, 1 + count(a.x) ORDER BY a.x % 2`
+* The sub-expression `a.x` is a recognized grouping key.
+* The sub-expression `2` is a constant.
+
+. `MATCH (a) WITH a.x AS ax RETURN ax, ax + count(ax) ORDER BY ax`
+* The expression `ax` is a recognized grouping key.
+
+. `MATCH (a)-[]->(b) RETURN a, a.x + count(b.y) ORDER BY a.y`
+* The sub-expression `a` is a recognized grouping key.
+
+. `MATCH (a)-[]->(b) RETURN a, count(b.y) ORDER BY size(keys(a))`
+* The sub-expression `a` is a recognized grouping key.
+
+. `MATCH (x) RETURN x.a, x.b, x.c, x.a + x.b + count(x) + x.c ORDER BY x.a + x.c`
+* The sub-expressions `x.a` and `x.c` are recognized grouping keys.
+
+. `MATCH (a) WITH a.x + 1 as ax RETURN ax, ax - 1 + count(ax) ORDER BY ax - 1`
+* The sub-expression `ax` is a recognized grouping key.
+* The sub-expression `1` is a constant.
+
+. `MATCH (a) WITH a.x + 1 as ax RETURN ax, ax - 1 + count(ax) ORDER BY ax + 2`
+* The sub-expression `ax` is a recognized grouping key.
+* The sub-expression `2` is a constant.
+
+. `MATCH (a) WITH a.x + 1 as ax RETURN ax AS x, ax - 1 + count(ax) ORDER BY x + 2`
+* The sub-expression `x` is an recognized alias.
+* The sub-expression `2` is a constant.
+
+. `MATCH (a) WITH a.x + 1 as ax RETURN ax AS x, ax - 1 + count(ax) AS y ORDER BY x + 2 - y`
+* The sub-expressions `x` and `y` are recognized aliases.
+* The sub-expression `2` is a constant.
+
+. `WITH {a:1, b:2} AS map RETURN map.a, map.a + count(map.b) ORDER BY map.a`
+* The expression `map.a` is a recognized grouping key.
+
+=== Invalid orderings
+
+The following clauses exhibit invalid row orderings.
+For each example we list why it is invalid.
+
+. `RETURN 1 + count(\*) ORDER BY 1 + count(*)`
+* The expression `1 + count(*)` is not a grouping key.
+
+. `MATCH (a) RETURN a.x + 1, a.x + 1 + count(a.x) ORDER BY a.x + 1`
+* The sub-expression `a.x + 1` is a recognized grouping key.
+
+. `MATCH (a) RETURN a.x + 1, a.x + 1 + count(a.x) ORDER BY a.x + 2`
```

Comment:
Just as a note. This (and some of the examples below) doesn't run at all. It gives an exception: `In a WITH/RETURN with DISTINCT or an aggregation, it is not possible to access variables declared before the WITH/RETURN: a `. But I believe that might be another discussion and not related to this  CIP.

Reply:
Well, it is listed under "Invalid orderings" so `MATCH (a) RETURN a.x + 1, a.x + 1 + count(a.x) ORDER BY a.x + 2` is supposed to fail. 

However, on the one above, `MATCH (a) RETURN a.x + 1, a.x + 1 + count(a.x) ORDER BY a.x + 1`: The explanation text misses a "not", I think. It should say "The sub-expression `a.x + 1` is **not** a recognized grouping key." This, I will have to fix.